docs/proposals/volume-hostpath-qualifiers.md

```diff
@@ -0,0 +1,96 @@
+
+
+## Proposed API Change
+
+### Host Volume
+
+I propose that the [`v1.HostPathVolumeSource`](https://github.com/kubernetes/kubernetes/blob/d26b4ca2859aa667ad520fb9518e0db67b74216a/pkg/api/types.go#L447-L451) object be changed to include the following additional field:
+
+`Type` - An optional string of `auto|exists|file|device|socket|directory` - If not set, defaults to `auto`. These values have the following behavior:
+
+| Value | Behavior |
+|:------|:---------|
+| `auto ` | If nothing exists at the given path, an empty directory will be created there. Otherwise, behaves like `exists` |
+| `exists` | If nothing exists at the given path, the pod will fail to run and provide an informative error message |
+| `file` | If a file does not exist at the given path, the pod will fail to run and provide an informative error message |
+| `device` | If a block or character device does not exist at the given path, the pod will fail to run and provide an informative error message |
+| `socket` | If a socket does not exist at the given path, the pod will fail to run and provide an informative error message |
+| `directory` | If a directory does not exist at the given path, the pod will fail to run and provide an informative error message |
+
+Additional possible values, which are proposed to be excluded:
+
+|Value | Behavior | Reason for exclusion |
+|:-----|:---------|:---------------------|
+| `new-directory` | Like `auto`, but the given path must be a directory if it exists | `auto` mostly fills this use-case |
+| `character-device` |  | Granularity beyond `device` is likely to not help |
+| `block-device` |  | Granularity beyond `device` is likely to not help |
+| `new-file` | Like file, but if nothing exist an empty file is created instead | In general, bindmounting the parent directory of the file you intend to create addresses this usecase |
+
+### Volume SubPaths
+
+Volumes may also have a `subPath` specified, which similarly may or may not exist. I similarly propose that the [v1.VolumeMount](https://github.com/kubernetes/kubernetes/blob/d26b4ca2859aa667ad520fb9518e0db67b74216a/pkg/api/types.go#L886-L897) type be augmented with a `SubPathType` field which the exact same semantics as above.
+While I propose exposing the same set of possible values for consistency, in reality a `SubPath` is useful primarily under a network mount. Because of this, the `device` and `socket` values would likely not be useful.
+
+I do not have any examples of concrete examples from the Kubernetes project for this, in part because there are very few examples of `SubPath` to begin with.
+However, each of the use-cases described for Host Volumes also could conceivably occur as subdirectories of a network mount using this feature, and thus still offer insight into why it would be of value.
```

Comment:
I'm aware of folks using it to share host NFS mounts in predictable ways across a cluster (I think)